src/std/sys_android.c

```diff
@@ -0,0 +1,242 @@
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include <hl.h>
+
+#if defined(HL_MOBILE) && defined(HL_ANDROID)
+
+#ifndef HL_ANDROID_ACTIVITY
+#	define HL_ANDROID_ACTIVITY "org/haxe/HashLinkActivity"
+#endif
+
+#include <jni.h>
+#include <android/log.h>
+#include <pthread.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#ifndef HL_JNI_LOG_TAG
+#define HL_JNI_LOG_TAG "HL_JNI"
+#endif
+
+#define HL_ANDROID_EXTERNAL_STORAGE_NOT_MOUNTED	(0)
+#define HL_ANDROID_EXTERNAL_STORAGE_MOUNTED_RO	(1)
+#define HL_ANDROID_EXTERNAL_STORAGE_MOUNTED_RW	(2 | HL_ANDROID_EXTERNAL_STORAGE_MOUNTED_RO)
+
+/* #define LOGI(...)  __android_log_print(ANDROID_LOG_INFO,HL_JNI_LOG_TAG,__VA_ARGS__) */
+/* #define LOGE(...)  __android_log_print(ANDROID_LOG_ERROR,HL_JNI_LOG_TAG,__VA_ARGS__) */
+#define LOGI(...) do {} while (0)
+#define LOGE(...) do {} while (0)
+
+/* Forward declaration for JNI thread management */
+static void hl_android_jni_thread_destructor(void*);
+
+/* pthread key for proper JVM thread handling */
+static pthread_key_t hl_java_thread_key;
+
+/* Java VM reference */
+static JavaVM* hl_java_vm;
+
+/* Main activity */
+static jclass hl_java_activity_class;
+
+/* Method signatures */
+static jmethodID hl_java_method_id_get_context;
+
+/* Paths */
+static char *hl_android_external_files_path = NULL;
+static char *hl_android_internal_files_path = NULL;
+
+/* Function to retrieve JNI environment, and dealing with threading */
+static JNIEnv* hl_android_jni_get_env(void)
+{
+	/* Always try to attach if calling from a non-attached thread */
+	JNIEnv *env;
+	if((*hl_java_vm)->AttachCurrentThread(hl_java_vm, &env, NULL) < 0) {
+		LOGE("failed to attach current thread");
+		return 0;
+	}
+	pthread_setspecific(hl_java_thread_key, (void*) env);
+
+	return env;
+}
+
+/* JNI_OnLoad is automatically called when loading shared library through System.loadLibrary() Java call */
+JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM* vm, void* reserved)
+{
+	JNIEnv *env;
+	jclass cls;
+
+	hl_java_vm = vm;
+	if ((*hl_java_vm)->GetEnv(hl_java_vm, (void**) &env, JNI_VERSION_1_4) != JNI_OK) {
```

Comment:
@rtissera why this is needed if we overwrite `env` anyway later in `hl_android_jni_get_env` call?

PS I'm just asking because I'm new to all this, so trying to wrap my head around the initialization procedure.